content/oci/deletableoci.go

```diff
@@ -0,0 +1,216 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package oci provides access to an OCI content store.
+// Reference: https://github.com/opencontainers/image-spec/blob/v1.1.0-rc4/image-layout.md
+package oci
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"oras.land/oras-go/v2/content"
+	"oras.land/oras-go/v2/internal/container/set"
+)
+
+// DeletableStore implements `oras.Target`, and represents a content store
+// extended with the delete operation.
+// Reference: https://github.com/opencontainers/image-spec/blob/v1.1.0-rc4/image-layout.md
+type DeletableStore struct {
+	*Store
+	lock sync.RWMutex
+}
+
+// NewDeletableStore returns a new DeletableStore with context.Background().
+func NewDeletableStore(root string) (*DeletableStore, error) {
+	store, err := NewWithContext(context.Background(), root)
+	if err != nil {
+		return nil, err
+	}
+	ds := &DeletableStore{
+		Store: store,
+	}
+	return ds, nil
+}
+
+// NewDeletableStoreWithContext returns a new DeletableStore.
+func NewDeletableStoreWithContext(ctx context.Context, root string) (*DeletableStore, error) {
+	store, err := NewWithContext(ctx, root)
+	if err != nil {
+		return nil, err
+	}
+	ds := &DeletableStore{
+		Store: store,
+	}
+	return ds, nil
+}
+
+// Fetch fetches the content identified by the descriptor.
+func (ds *DeletableStore) Fetch(ctx context.Context, target ocispec.Descriptor) (io.ReadCloser, error) {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
+
+	return ds.Store.Fetch(ctx, target)
+}
+
+// Push pushes the content, matching the expected descriptor.
```

Comment:
You need to document the perf loss (e.g. cannot push concurrently) here

Reply:
Added the doc comment.

Reply:
I realize that we only need to avoid `Delete` and `Push` happening at the same time. Maybe we can refine the locks to allow concurrent Push. Anyway save performance improvements to later for now.